File: code/richediting/src/main/java/org/fbme/ide/richediting/adapters/fb/FBTypeTemplateCellComponent.java
package org.fbme.ide.richediting.adapters.fb;

import jetbrains.mps.editor.runtime.TextBuilderImpl;
import jetbrains.mps.editor.runtime.style.StyleAttributes;
import jetbrains.mps.nodeEditor.MPSColors;
import jetbrains.mps.nodeEditor.cellLayout.AbstractCellLayout;
import jetbrains.mps.nodeEditor.cells.EditorCell_Collection;
import jetbrains.mps.nodeEditor.cells.ParentSettings;
import jetbrains.mps.openapi.editor.EditorContext;
import jetbrains.mps.openapi.editor.TextBuilder;
import jetbrains.mps.openapi.editor.cells.EditorCell;
import org.fbme.lib.iec61499.descriptors.FBTypeDescriptor;
import org.fbme.lib.iec61499.fbnetwork.EntryKind;
import org.fbme.scenes.controllers.LayoutUtil;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.mps.openapi.model.SNode;

import java.awt.*;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class FBTypeTemplateCellComponent {

    private final FBTypeDescriptor myFBType;
    private final FBTypeCellComponent myFBCellComponent;
    private final EditorCell_Collection myCellCollection;
    private final List<EventAssociation> myInputAssociations = new ArrayList<>();
    private final List<EventAssociation> myOutputAssociations = new ArrayList<>();

    public FBTypeTemplateCellComponent(@NotNull EditorContext editorContext, SNode node, FBTypeDescriptor fbType) {
        myFBType = fbType;
        myCellCollection = createRootCell(editorContext, node);
        myFBCellComponent = new FBTypeCellComponent(editorContext, fbType, node, true);
        myCellCollection.addEditorCell(myFBCellComponent.getRootCell());
        init();
    }


    public EditorCell_Collection getRootCell() {
        return myCellCollection;
    }

    private EditorCell_Collection createRootCell(@NotNull EditorContext editorContext, SNode node) {
        return new EditorCell_Collection(editorContext, node, new AbstractCellLayout() {
            public void doLayout(jetbrains.mps.openapi.editor.cells.EditorCell_Collection cells) {
                assert cells == myCellCollection;
                FBTypeTemplateCellComponent.this.relayout();
            }

            public TextBuilder doLayoutText(Iterable<EditorCell> cells) {
                return new TextBuilderImpl();
            }
        }) {
            @Override
            protected void paintContent(Graphics graphics, ParentSettings settings) {
                graphics = graphics.create();
                graphics.setColor(getForegroundColor());

                int x = getX();
                int y = getY();

                EventsAssociationsPainter inputAssociationPainter = new EventsAssociationsPainter(myInputAssociations);
                EventsAssociationsPainter outputAssociationPainter = new EventsAssociationsPainter(myOutputAssociations);

                inputAssociationPainter.paint(graphics, true, x, y);
                x += inputAssociationPainter.getWidth();
                x += myFBCellComponent.getWidth();
                outputAssociationPainter.paint(graphics, false, x, y);
            }
        };
    }

    public void relayout() {
        EditorCell_Collection rootCell = myFBCellComponent.getRootCell();
        rootCell.relayout();

        int inputsWidth = new EventsAssociationsPainter(myInputAssociations).getWidth();
        int outputsWidth = new EventsAssociationsPainter(myOutputAssociations).getWidth();

        rootCell.moveTo(myCellCollection.getX() + inputsWidth, myCellCollection.getY());

        myCellCollection.setWidth(inputsWidth + myFBCellComponent.getWidth() + outputsWidth);
        myCellCollection.setHeight(myFBCellComponent.getHeight());
    }

    public void init() {
        for (int inputEvent = 0; inputEvent < myFBType.getEventInputPorts().size(); inputEvent++) {
            List<Integer> inputVariables = myFBType.getAssociatedVariablesForInputEvent(inputEvent);
            if (!inputVariables.isEmpty()) {
                ArrayList<Integer> sortedInputVariables = new ArrayList<>(inputVariables);
                Collections.sort(sortedInputVariables);
                myInputAssociations.add(new EventAssociation(inputEvent, sortedInputVariables));
            }
        }
        for (int outputEvent = 0; outputEvent < myFBType.getEventOutputPorts().size(); outputEvent++) {
            List<Integer> outputVariables = myFBType.getAssociatedVariablesForOutputEvent(outputEvent);
            if (!outputVariables.isEmpty()) {
                ArrayList<Integer> sortedOutputVariables = new ArrayList<>(outputVariables);
                Collections.sort(sortedOutputVariables);
                myOutputAssociations.add(new EventAssociation(outputEvent, sortedOutputVariables));
            }
        }

        myCellCollection.setHeight(myFBCellComponent.getHeight());

        int inputAssocationsWidth = new EventsAssociationsPainter(myInputAssociations).getWidth();
        int outputAssocationsWidth = new EventsAssociationsPainter(myOutputAssociations).getWidth();
        myCellCollection.setWidth(inputAssocationsWidth + myFBCellComponent.getWidth() + outputAssocationsWidth);
    }

    public static class EventAssociation {
        private final int myEventNumber;
        private final List<Integer> myVariableNumbers;

        public EventAssociation(int eventNumber, List<Integer> variableNumbers) {
            myEventNumber = eventNumber;
            myVariableNumbers = variableNumbers;
        }
    }

    private static final int GAP_SPACES = 2;

    private class EventsAssociationsPainter {
        private final List<EventAssociation> myAsociations;
        private final int myLineSize = LayoutUtil.getLineSize(myCellCollection.getStyle());
        private final int myGapWidth = GAP_SPACES * myLineSize;
        private final int myAssociationsWidth;

        private EventsAssociationsPainter(List<EventAssociation> eventAsociations) {
            myAsociations = eventAsociations;
            myAssociationsWidth = eventAsociations.size() * myLineSize;
        }

        private int getWidth() {
            return myGapWidth + myAssociationsWidth;
        }

        private void paint(Graphics graphics, boolean inputPorts, int x, int y) {
            paintLines(graphics, inputPorts, x, y);
            paintAssociations(graphics, inputPorts, x, y);
        }

        private void paintLines(Graphics graphics, boolean inputPorts, int x, int y) {
            for (int i = 0; i < getEventPortsCount(inputPorts); i++) {
                int portY = getEventPortPosition(inputPorts, i, y);
                graphics.drawLine(x, portY, x + getWidth(), portY);
            }
            for (int i = 0; i < getBottomPortsCount(inputPorts); i++) {
                int portY = getDataPortPosition(inputPorts, i, y);
                graphics.drawLine(x, portY, x + getWidth(), portY);
            }
        }

        private void paintAssociations(Graphics graphics, boolean inputPorts, int x, int y) {
            x += myGapWidth / 2;
            x += myLineSize / 2;
            for (EventAssociation association : myAsociations) {
                int eventNumber = association.myEventNumber;
                List<Integer> dataNumbers = association.myVariableNumbers;
                int lastVariableNumber = dataNumbers.get(dataNumbers.size() - 1);
                int eventY = getEventPortPosition(inputPorts, eventNumber, y);
                graphics.drawLine(x, eventY, x, getDataPortPosition(inputPorts, lastVariableNumber, y));

                graphics.setColor(DiagramColors.getColorFor(EntryKind.EVENT));
                graphics.fillRect(x - 2, eventY - 2, 4, 4);
                graphics.setColor(getForegroundColor());
                graphics.drawRect(x - 2, eventY - 2, 4, 4);

                for (Integer dataNumber : dataNumbers) {
                    int dataY = getDataPortPosition(inputPorts, dataNumber, y);
                    graphics.setColor(DiagramColors.getColorFor(EntryKind.DATA));
                    graphics.fillRect(x - 2, dataY - 2, 4, 4);
                    graphics.setColor(getForegroundColor());
                    graphics.drawRect(x - 2, dataY - 2, 4, 4);
                }
                x += myLineSize;
            }
        }

        private int getEventPortsCount(boolean input) {
            return input ? myFBCellComponent.getInputEventPortsCount() : myFBCellComponent.getOutputEventPortsCount();
        }

        private int getEventPortPosition(boolean input, int number, int offsetY) {
            return (input ? myFBCellComponent.getInputEventPortPosition(number) : myFBCellComponent.getOutputEventPortPosition(number)).y + offsetY;
        }

        private int getBottomPortsCount(boolean input) {
            return input ? myFBCellComponent.getInputBottomPortsCount() : myFBCellComponent.getOutputBootomPortsCount();
        }

        private int getDataPortPosition(boolean input, int number, int offsetY) {
            return (input ? myFBCellComponent.getInputDataPortPosition(number) : myFBCellComponent.getOutputDataPortPosition(number)).y + offsetY;
        }
    }

    private Color getForegroundColor() {
        Color foreground = myCellCollection.getStyle().get(StyleAttributes.TEXT_COLOR);
        return foreground == null ? MPSColors.BLACK : foreground;
    }

}

File: code/richediting/src/main/kotlin/org/fbme/ide/richediting/adapters/fbnetwork/fb/FBTypeTemplateCellComponent.kt
package org.fbme.ide.richediting.adapters.fbnetwork.fb

import jetbrains.mps.editor.runtime.TextBuilderImpl
import jetbrains.mps.editor.runtime.style.StyleAttributes
import jetbrains.mps.nodeEditor.MPSColors
import jetbrains.mps.nodeEditor.cellLayout.AbstractCellLayout
import jetbrains.mps.nodeEditor.cells.EditorCell_Collection
import jetbrains.mps.nodeEditor.cells.ParentSettings
import jetbrains.mps.openapi.editor.EditorContext
import jetbrains.mps.openapi.editor.TextBuilder
import jetbrains.mps.openapi.editor.cells.EditorCell
import org.fbme.lib.iec61499.descriptors.FBTypeDescriptor
import org.fbme.lib.iec61499.fbnetwork.EntryKind
import org.fbme.scenes.controllers.LayoutUtil.getLineSize
import org.jetbrains.mps.openapi.model.SNode
import java.awt.Color
import java.awt.Graphics

class FBTypeTemplateCellComponent(editorContext: EditorContext, node: SNode, private val myFBType: FBTypeDescriptor) {
    private val myFBCellComponent: FBTypeCellComponent
    val cellCollection: EditorCell_Collection
    private val myInputAssociations: MutableList<EventAssociation> = ArrayList()
    private val myOutputAssociations: MutableList<EventAssociation> = ArrayList()

    private fun createRootCell(editorContext: EditorContext, node: SNode): EditorCell_Collection {
        return object : EditorCell_Collection(
            editorContext,
            node,
            object : AbstractCellLayout() {
                override fun doLayout(cells: jetbrains.mps.openapi.editor.cells.EditorCell_Collection) {
                    assert(cells === cellCollection)
                    relayout()
                }

                override fun doLayoutText(cells: Iterable<EditorCell>): TextBuilder {
                    return TextBuilderImpl()
                }
            }
        ) {
            override fun paintContent(graphics: Graphics, settings: ParentSettings) {
                var g = graphics
                g = g.create()
                g.color = foregroundColor
                var x = x
                val y = y
                val inputAssociationPainter = EventsAssociationsPainter(myInputAssociations)
                val outputAssociationPainter = EventsAssociationsPainter(myOutputAssociations)
                inputAssociationPainter.paint(g, true, x, y)
                x += inputAssociationPainter.width
                x += myFBCellComponent.width
                outputAssociationPainter.paint(g, false, x, y)
            }
        }
    }

    fun relayout() {
        val rootCell = myFBCellComponent.rootCell
        rootCell.relayout()
        val inputsWidth: Int = EventsAssociationsPainter(myInputAssociations).width
        val outputsWidth: Int = EventsAssociationsPainter(myOutputAssociations).width
        rootCell.moveTo(cellCollection.x + inputsWidth, cellCollection.y)
        cellCollection.width = inputsWidth + myFBCellComponent.width + outputsWidth
        cellCollection.height = myFBCellComponent.height
    }

    fun init() {
        for (inputEvent in myFBType.eventInputPorts.indices) {
            val inputVariables = myFBType.getAssociatedVariablesForInputEvent(inputEvent)
            if (inputVariables.isNotEmpty()) {
                val sortedInputVariables = inputVariables.sorted()
                myInputAssociations.add(EventAssociation(inputEvent, sortedInputVariables))
            }
        }
        for (outputEvent in myFBType.eventOutputPorts.indices) {
            val outputVariables = myFBType.getAssociatedVariablesForOutputEvent(outputEvent)
            if (outputVariables.isNotEmpty()) {
                val sortedOutputVariables = outputVariables.sorted()
                myOutputAssociations.add(EventAssociation(outputEvent, sortedOutputVariables))
            }
        }
        cellCollection.height = myFBCellComponent.height
        val inputAssociationsWidth: Int = EventsAssociationsPainter(myInputAssociations).width
        val outputAssociationsWidth: Int = EventsAssociationsPainter(myOutputAssociations).width
        cellCollection.width = inputAssociationsWidth + myFBCellComponent.width + outputAssociationsWidth
    }

    class EventAssociation(val eventNumber: Int, val variableNumbers: List<Int>)

    private inner class EventsAssociationsPainter(private val associations: List<EventAssociation>) {
        private val myLineSize = getLineSize(cellCollection.style)
        private val myGapWidth = GAP_SPACES * myLineSize
        private val myAssociationsWidth: Int = associations.size * myLineSize
        val width: Int
            get() = myGapWidth + myAssociationsWidth

        fun paint(graphics: Graphics, inputPorts: Boolean, x: Int, y: Int) {
            paintLines(graphics, inputPorts, x, y)
            paintAssociations(graphics, inputPorts, x, y)
        }

        private fun paintLines(graphics: Graphics, inputPorts: Boolean, x: Int, y: Int) {
            for (i in 0 until getEventPortsCount(inputPorts)) {
                val portY = getEventPortPosition(inputPorts, i, y)
                graphics.drawLine(x, portY, x + width, portY)
            }
            for (i in 0 until getBottomPortsCount(inputPorts)) {
                val portY = getDataPortPosition(inputPorts, i, y)
                graphics.drawLine(x, portY, x + width, portY)
            }
        }

        private fun paintAssociations(graphics: Graphics, inputPorts: Boolean, x: Int, y: Int) {
            var curX = x
            curX += myGapWidth / 2
            curX += myLineSize / 2
            for (association in associations) {
                val eventNumber = association.eventNumber
                val dataNumbers = association.variableNumbers
                val lastVariableNumber = dataNumbers[dataNumbers.size - 1]
                val eventY = getEventPortPosition(inputPorts, eventNumber, y)
                graphics.drawLine(curX, eventY, curX, getDataPortPosition(inputPorts, lastVariableNumber, y))
                graphics.color = DiagramColors.getColorFor(EntryKind.EVENT)
                graphics.fillRect(curX - 2, eventY - 2, 4, 4)
                graphics.color = foregroundColor
                graphics.drawRect(curX - 2, eventY - 2, 4, 4)
                for (dataNumber in dataNumbers) {
                    val dataY = getDataPortPosition(inputPorts, dataNumber, y)
                    graphics.color = DiagramColors.getColorFor(EntryKind.DATA)
                    graphics.fillRect(curX - 2, dataY - 2, 4, 4)
                    graphics.color = foregroundColor
                    graphics.drawRect(curX - 2, dataY - 2, 4, 4)
                }
                curX += myLineSize
            }
        }

        private fun getEventPortsCount(input: Boolean): Int {
            return if (input) myFBCellComponent.inputEventPortsCount else myFBCellComponent.outputEventPortsCount
        }

        private fun getEventPortPosition(input: Boolean, number: Int, offsetY: Int): Int {
            val portPosition = if (input) {
                myFBCellComponent.getInputEventPortPosition(number)
            } else {
                myFBCellComponent.getOutputEventPortPosition(number)
            }
            return portPosition.y + offsetY
        }

        private fun getBottomPortsCount(input: Boolean): Int {
            return if (input) myFBCellComponent.inputBottomPortsCount else myFBCellComponent.outputBottomPortsCount
        }

        private fun getDataPortPosition(input: Boolean, number: Int, offsetY: Int): Int {
            val portPosition = if (input) {
                if (number >= myFBCellComponent.inputDataPortsCount) {
                    myFBCellComponent.getSocketPortPosition(number - myFBCellComponent.inputDataPortsCount)
                } else {
                    myFBCellComponent.getInputDataPortPosition(number)
                }
            } else {
                if (number >= myFBCellComponent.outputDataPortsCount) {
                    myFBCellComponent.getPlugPortPosition(number - myFBCellComponent.outputDataPortsCount)
                } else {
                    myFBCellComponent.getOutputDataPortPosition(number)
                }
            }
            return portPosition.y + offsetY
        }
    }

    private val foregroundColor: Color
        get() {
            val foreground = cellCollection.style.get(StyleAttributes.TEXT_COLOR)
            return foreground ?: MPSColors.BLACK
        }

    companion object {
        private const val GAP_SPACES = 2
    }

    init {
        cellCollection = createRootCell(editorContext, node)
        myFBCellComponent = FBTypeCellComponent(editorContext, myFBType, node, true)
        cellCollection.addEditorCell(myFBCellComponent.rootCell)
        init()
    }
}
